File: discipline_daemon_lib/src/conditionals/countdown_conditional.rs
use serde::{Deserialize, Serialize};
use crate::x::{Countdown, Duration, Instant};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CountdownConditional {
  pub duration: Duration,
  pub countdown: Option<Countdown>,
}

impl CountdownConditional {
  pub fn create(duration: Duration) -> CountdownConditional {
    CountdownConditional { duration, countdown: None }
  }
  
  pub fn is_active(&self, now: Instant) -> bool {
    matches!(&self.countdown, Some(countdown) if countdown.is_running(now))
  }

  pub fn activate(&mut self, now: Instant) {
    self.countdown = Some(Countdown::construct(now, self.duration));
  }

  pub fn create_activate_state(&self, now: Instant) -> CountdownConditionalActivateState {
    CountdownConditionalActivateState { countdown: Countdown::construct(now, self.duration) }
  }

  pub fn activate_from_activate_state(&mut self, activate_state: CountdownConditionalActivateState) {
    self.countdown = Some(activate_state.countdown);
  }
}

pub struct CountdownConditionalActivateState {
  pub countdown: Countdown,
}

File: discipline_kt/app/src/main/java/com/example/app/conditionals/CountdownConditional.kt
package com.example.app

public data class CountdownConditional(
  var duration: Duration,
  var countdown: Countdown?
) {
  companion object {
    fun create(duration: Duration): CountdownConditional {
      return CountdownConditional(duration, null)
    }

    fun construct(duration: Duration, countdown: Countdown?): CountdownConditional {
      return CountdownConditional(duration, countdown)
    }
  }

  fun isActive(now: Instant): Boolean {
    val countdown = countdown ?: return false
    return countdown.isRunning(now)
  }

  fun reactivate(now: Instant) {
    countdown = Countdown.create(now, duration)
  }

  class ReactivateState(val countdown: Countdown) {}

  fun createReactivateState(now: Instant): ReactivateState {
    return ReactivateState(Countdown.create(now, duration))
  }

  fun reactivateFromState(reactivateState: ReactivateState) {
    countdown = reactivateState.countdown
  }
}